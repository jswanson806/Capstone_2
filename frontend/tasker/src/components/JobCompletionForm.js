import React, { useState } from "react";
import TaskerApi from "../api";
import { Button, Input, Form, FormGroup, Label, FormFeedback, FormText } from "reactstrap";
import "./styles/JobCompletionForm.css";

const JobCompletionForm = ({job, onUpload}) => {    

    const [imageFile, setImageFile] = useState('');
    const [isValid, setIsValid] = useState(true);
    
    // default rate for Tasker -> (once reviews are implemented, workers should have variable rates depending on ratings)
    const rate = 20.00;
    // array of valid file types for uploading an image
    const validFileTypes = ['image/jpg', 'image/jpeg', 'image/png'];

    /** Handles form submission
     * 
     * Creates job update data and appends imageFile to the form
     * 
     * calls API to update job and upload after image to AWS S3
     * 
     * Calls onUpload function recieved from JobCard component
     */
    const handleSubmit = async (e) => {
        e.preventDefault();
       
        try{
            const updateInfo = createJobUpdate();
            // create job update object and spread job
            const jobInfo = { job: {...updateInfo}};
           
            // initilize new FormData object and append image/imageFile key/value pair
            const form = new FormData();
            form.append('image', imageFile);
            
            // update job in db
            await TaskerApi.updateSingleJob(jobInfo);
            // call api to upload file to S3 storage
            await TaskerApi.uploadAfterImage(form, job.assigned_to);
            // call onUpload function from JobCard component, which calls endWork to hide the form and trigger rerender
            onUpload();
        } catch(err) {
            console.error('Error:', err);
        }
        
    };

    /** Handles file upload
     * 
     * Varifies file type is valid
     * 
     * Updates state of isValid and imageFile
     */
    const handleUpload = async (e) => {
        // get the target file
        const file = e.target.files[0];
        // if filetype not in array of valid files, update state and return
        if(!validFileTypes.find(type => type === file.type)){
            setIsValid(false);
            return;
        };
        // update state of isValid
        setIsValid(true);
        // update state of imageFile
        setImageFile(file);

    };

    /** Creates the job update data object
     * 
     * Updates job properties with values for after_image_url, end_time, payment_due, and status
     * 
     * Calls caluclatedHoursWorked and calculatePaymentDue
     * 
     * returns job object
     */

    const createJobUpdate = () => {

        // set options for toLocalString method of the Date object
        const options = {
            year: 'numeric',
            month: 'numeric',
            day: 'numeric',
            hour: 'numeric', 
            minute: 'numeric', 
            second: 'numeric'
        }
        // initialize new Date object
        const currentTime = new Date();
        // variable will hold calculated hours worked
        let hours;

        // update the end_time property to current time/date
        job.end_time = currentTime.toLocaleString(undefined, options);
        // with end_time set, calculate hours worked
        hours = calculateHoursWorked(job);
        // with calculated hours, calculate payment due
        job.payment_due = calculatePaymentDue(hours, rate);
        // update the after_image_url property to the value of image form field
        job.after_image_url = imageFile.name;
        // update job status
        job.status = 'pending review';
 
        return job;
    }

    /** Calculate the time spent working on the job in hours
     * 
     * Returns hours worked
     */
    const calculateHoursWorked = (job) => {
        // initialize new Date objects for start and end times
        const startDate = new Date(job.start_time);
        const endDate = new Date(job.end_time);

        // Calculate the time difference in milliseconds
        const timeDifferenceMs = endDate - startDate;

        // Calculate the time difference in hours
        const hoursWorked = timeDifferenceMs / (1000 * 60 * 60);

        // return time in hours
        return hoursWorked;
    }

    /** Calculate the payment due for the job in cents (smallest currency unit for Stripe API)
     * 
     * Returns payment due as integer
     */

    const calculatePaymentDue = (hours, rate) => {
        // calculate payment due in cents based on hours worked
        const paymentInCents = Math.round(hours * rate * 100);
        //return payment due in cents
        return paymentInCents;
    }


    return (
        <div className="jobCompletion-form-container">
            <div className="jobCompletion-centered-card">
                {/* form for completing job - calls handleUpload and handleSubmit*/}
                <Form onSubmit={handleSubmit}>
                    <FormGroup>
                      <Input
                        id="imageInput"
                        type="file"
                        name="image"
                        hidden
                        onChange={handleUpload}
                    />
                    </FormGroup>
                    <FormGroup>
                        <Label for="imageInput">
                            Upload "After" Image
                        </Label>
                        <Input
                            id="imageInput"
                            type="file"
                            name="image"
                            data-testid="image-input"
                            onChange={handleUpload}
                            valid={isValid}
                            invalid={!isValid}
                        />
                        {/* provide form feedback for iamge upload based on isValid state */}
                        {!isValid 
                            ? 
                            <FormFeedback >Invalid file type</FormFeedback> 
                            : 
                            <FormFeedback valid>Valid file type</FormFeedback>
                        }
                        <FormText>Supported File Types: jpg, png</FormText>
                    </FormGroup>
                    <Button type="submit" data-testid="jobCompletion-submit-button">Upload</Button>
                </Form>
            </div>
        </div>
    )
}

export default JobCompletionForm;