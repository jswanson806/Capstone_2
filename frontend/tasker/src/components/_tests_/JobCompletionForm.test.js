import React from "react";
import JobCompletionForm from "../JobCompletionForm";
import {act, render, screen, fireEvent, waitFor} from '@testing-library/react';
import TaskerApi from "../../api";

const job = {
    id: 1, 
    title: 'testing jobs 1', 
    body: 'job 1 body for testing the job card', 
    postedBy: 1, 
    status: 'active',
    start_time: '2023-10-03 09:00:00 AM',
    applicants: []
}

const onUpload = jest.fn();
const updateSingleJob = jest.spyOn(TaskerApi, 'updateSingleJob');
const uploadAfterImage = jest.spyOn(TaskerApi, 'uploadAfterImage');

describe("JobCompletionForm smoke and snapshot tests", () => {

    it("should render without crashing", async () => {

        await act(async () => {render(
            <JobCompletionForm job={job} onUpload={onUpload}/>
        )

        });
    });

    it("should match snapshot", async() => {
        let asFragment;

        await act(async () => {const {asFragment: fragment} = render(
            <JobCompletionForm job={job} onUpload={onUpload}/>
        )
        asFragment = fragment;
        });

        expect(asFragment()).toMatchSnapshot();
    })
})


describe("handles form submission correctly", () => {

    it("should render without crashing", async () => {

        await act(async () => {render(
            <JobCompletionForm job={job} onUpload={onUpload}/>
        )

        });
    });

    it("should match snapshot", async() => {
        let asFragment;

        await act(async () => {const {asFragment: fragment} = render(
            <JobCompletionForm job={job} onUpload={onUpload}/>
        )
        asFragment = fragment;
        });

        expect(asFragment()).toMatchSnapshot();
    });
});


describe("Handles file upload correctly", () => {

    it('handles valid file upload', async () => {
        await act(async () => {render(
            <JobCompletionForm job={job} onUpload={onUpload}/>
        )
        });

        const inputElement = screen.getByTestId('image-input');
      
        // fake file
        const fakeFile = new File(['fake image content'], 'fake-image.png', {
          type: 'image/png',
        });
      
        // Simulate a file change event
        fireEvent.change(inputElement, { target: { files: [fakeFile] } });
      
        // input is valid 
        expect(inputElement).toBeValid();

    });

    it('handles invalid file upload', async () => {
        await act(async () => {render(
            <JobCompletionForm job={job} onUpload={onUpload}/>
        )
        });

        const inputElement = screen.getByTestId('image-input');
      
        // fake file
        const fakeFile = new File(['fake video content'], 'fake-video.mpg', {
          type: 'video/mpg',
        });
      
        // Simulate a file change event
        fireEvent.change(inputElement, { target: { files: [fakeFile] } });
      
        // input is not valid 
        expect(inputElement).not.toBeValid();

    });

});

describe("Handles form submission correctly", () => {

    it('should create job update and call api upon submit with correct data', async () => {
        await act(async () => {render(
            <JobCompletionForm job={job} onUpload={onUpload}/>
        )
        });

        const inputElement = screen.getByTestId('image-input');
      
        // fake file
        const fakeFile = new File(['fake image content'], 'fake-image.png', {
          type: 'image/png',
        });
      
        // Simulate a file change event
        fireEvent.change(inputElement, { target: { files: [fakeFile] } });
      
        // input is valid 
        expect(inputElement).toBeValid();

        const submitBtn = screen.getByTestId("jobCompletion-submit-button");

        fireEvent.click(submitBtn)

        expect(updateSingleJob).toHaveBeenCalledTimes(1);
        expect(updateSingleJob).toHaveBeenCalledWith({ job: 
            {
                id: 1, 
                title: 'testing jobs 1', 
                body: 'job 1 body for testing the job card', 
                postedBy: 1, 
                status: 'pending review',
                start_time: '2023-10-03 09:00:00 AM',
                end_time: expect.any(String),
                after_image_url: 'fake-image.png',
                payment_due: expect.any(Number),
                applicants: []
            }
        });

    });

    it('should call api to upload after image upon submit', async () => {
        await act(async () => {render(
            <JobCompletionForm job={job} onUpload={onUpload}/>
        )
        });

        const inputElement = screen.getByTestId('image-input');
      
        // fake file
        const fakeFile = new File(['fake image content'], 'fake-image.png', {
          type: 'image/png',
        });
      
        // Simulate a file change event
        fireEvent.change(inputElement, { target: { files: [fakeFile] } });
      
        // input is valid 
        expect(inputElement).toBeValid();

        const submitBtn = screen.getByTestId("jobCompletion-submit-button");

        fireEvent.click(submitBtn)

        await waitFor(() => {
            expect(uploadAfterImage).toHaveBeenCalledTimes(1);
        })
        

    });

    it('should call onUpload upon submit', async () => {
        await act(async () => {render(
            <JobCompletionForm job={job} onUpload={onUpload}/>
        )
        });

        const inputElement = screen.getByTestId('image-input');
      
        // fake file
        const fakeFile = new File(['fake image content'], 'fake-image.png', {
          type: 'image/png',
        });
      
        // Simulate a file change event
        fireEvent.change(inputElement, { target: { files: [fakeFile] } });
      
        // input is valid 
        expect(inputElement).toBeValid();

        const submitBtn = screen.getByTestId("jobCompletion-submit-button");

        fireEvent.click(submitBtn)

        await waitFor(() => {
            expect(onUpload).toHaveBeenCalledTimes(1);
        })
        

    });


});