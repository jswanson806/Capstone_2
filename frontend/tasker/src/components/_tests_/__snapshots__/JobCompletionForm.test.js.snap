// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JobCompletionForm smoke and snapshot tests should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="jobCompletion-form-container"
  >
    <div
      class="jobCompletion-centered-card"
    >
      <form
        class=""
      >
        <div
          class="form-group"
        >
          <input
            class="form-control-file"
            hidden=""
            id="imageInput"
            name="image"
            type="file"
          />
        </div>
        <div
          class="form-group"
        >
          <label
            class=""
            for="imageInput"
          >
            Upload "After" Image
          </label>
          <input
            aria-invalid="false"
            class="is-valid form-control-file"
            data-testid="image-input"
            id="imageInput"
            name="image"
            type="file"
          />
          <div
            class="valid-feedback"
          >
            Valid file type
          </div>
          <small
            class="form-text text-muted"
          >
            Supported File Types: jpg, png
          </small>
        </div>
        <button
          class="btn btn-secondary"
          data-testid="jobCompletion-submit-button"
          type="submit"
        >
          Upload
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;

exports[`handles form submission correctly should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="jobCompletion-form-container"
  >
    <div
      class="jobCompletion-centered-card"
    >
      <form
        class=""
      >
        <div
          class="form-group"
        >
          <input
            class="form-control-file"
            hidden=""
            id="imageInput"
            name="image"
            type="file"
          />
        </div>
        <div
          class="form-group"
        >
          <label
            class=""
            for="imageInput"
          >
            Upload "After" Image
          </label>
          <input
            aria-invalid="false"
            class="is-valid form-control-file"
            data-testid="image-input"
            id="imageInput"
            name="image"
            type="file"
          />
          <div
            class="valid-feedback"
          >
            Valid file type
          </div>
          <small
            class="form-text text-muted"
          >
            Supported File Types: jpg, png
          </small>
        </div>
        <button
          class="btn btn-secondary"
          data-testid="jobCompletion-submit-button"
          type="submit"
        >
          Upload
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
